### PR TITLE
Deployment clone action and networking slowness injection

### PR DESCRIPTION
**Pull request by seeker89**
## TL;DR

This adds a new action - cloning a deployment to add failure into a copy, rather than modifying the existing software. It also lays foundation for various mutators, with a first batch of:

- environment variables mutation (for example override `REDIS_PORT` to a proxied one)
- [traffic control](https://tldp.org/HOWTO/Traffic-Control-HOWTO/intro.html) mutation (add latency to the whole pod)
- [toxiproxy](http://toxiproxy.io/) mutation with automatic redirection via `iptables` to selectively create proxies and configure network issues on them

That gives the user two different ways of achieving safe-ish (only the copy is affected) network slowness simulation (L3).

## Details

This is still a little rough around the edges and various things should be made configurable, but the concept is proven.

- [x] add the schema changes for the new clone action
- [x] implement the cloning mechanism for deployments
- [x] implement the label modifications
- [x] implement the mutations of environment variables
- [x] implement copying labels from services' selectors
- [x] implement traffic control mutation
- [ ] add tc presets like latency, loss, throttle
- [x] add traffic control docs and example usage
- [x] implement toxiproxy mutation
- [x] add toxiproxy docs and example usage
- [x] implement automatic cleanup
- [ ] test coverage
- [x] upgrade kubernetes version to `12.0.10a1` to leverage Kubernetes 1.16

Also note, that because toxiproxy is a little silly and allows for populating the proxies at startup, but not the toxics, I'm abusing the startup probe a little bit, to actually populate the toxiproxy with the user-requested stuff.

## Testing

See the added docs for example usage. You will be able to see these pods with network slowness automatically created for you:

image

image

image

Resolves https://github.com/powerfulseal/powerfulseal/issues/285

### CODE REVIEW

**Comment by seeker89** [comment]
Looking for feedback and reviews on this one :)

File: powerfulseal/policy/action_clone.py [hunk 1]
```diff
@@ -0,0 +1,331 @@
...
+      Modifies labels and deployment selectors as required by the schema
+    """
+    for label_modifier in self.schema.get("labels", []):
+
+      # handle the service selector lookup
+      spec = label_modifier.get("service")
+      if spec is not None:
+        try:
+          # get the service we're after
+          service = self.k8s_inventory.k8s_client.get_service(
```
Comment (altoic) [inline powerfulseal/policy/action_clone.py:81]: why are you using the labels from the service? is it not safe to assume the labels can be cloned directly like the [static below.](https://github.com/powerfulseal/powerfulseal/pull/302/files#diff-06d2b9d696f1d59448f5bda0d1777f7eR92-R97) i think you can still append new labels with out needing to a get labels from the service.

File: powerfulseal/policy/action_clone.py [hunk 2]
```diff
@@ -0,0 +1,331 @@
...
+      name=source_deployment.get("name"),
+      namespace=source_deployment.get("namespace"),
+    )
+    return deployment
+
+  def modify_labels(self, body, update_labels):
+    """
+      Modifies labels and deployment selectors as required by the schema
+    """
+    for label_modifier in self.schema.get("labels", []):
```
Comment (altoic) [inline powerfulseal/policy/action_clone.py:74]: why are you iterating on the labels? can you not update the labels after consolidating all the labels so you dont have to make multiple calls to get service.

File: powerfulseal/policy/action_clone.py [hunk 3]
```diff
@@ -0,0 +1,331 @@
...
+        ),
+      )
+    )
+
+    # precompute the iptables command
+    def get_port(listen_string):
+      parts = listen_string.split(":")
+      if len(parts) == 2:
+        return parts[-1]
+      return "80"
```
Comment (altoic) [inline powerfulseal/policy/action_clone.py:287]: if the default is 80 this should be added explicitly somewhere in the docs. otherwise it seems that if a listening port is not defined it defaults to 80 for the redirect.

File: powerfulseal/policy/action_clone.py (hunk 1 above)
Comment (seeker89) [reply powerfulseal/policy/action_clone.py:81]: What's not safe about it? This is an optional way of cloning a service selectors as labels. Very useful, when all you care about is that your new deployment is picked up by the service you're targeting.

File: powerfulseal/policy/action_clone.py (hunk 3 above)
Comment (seeker89) [reply powerfulseal/policy/action_clone.py:287]: Yup, this is a proof of concept. It should probably land in the docs, but defaulting to 80 is pretty standard.

File: powerfulseal/policy/action_clone.py (hunk 2 above)
Comment (seeker89) [reply powerfulseal/policy/action_clone.py:74]: I'm not sure what that sentence means. I'm not iterating on labels... I'm iterating on the label modifiers. You can specify multiple label modifiers, that can potentially override bits of each other. Example: copy labels over from service X matchers, and then add static labels of my choosing.
